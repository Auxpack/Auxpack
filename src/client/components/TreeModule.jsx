import React from 'react'

const TreeModule = (props) => {
  <div>
    <h3>Mixed Modules</h3>
    <p>{`Count: ${(bothCount !== 0) ? bothCount : 0}`}</p>
    <p>{`Percentage: ${Math.round(bothCount / totalCount * 100)}%`}</p>
  </div>
}

export default TreeModule;