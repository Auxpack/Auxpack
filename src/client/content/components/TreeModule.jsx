import React from 'react';
import Card from '@material-ui/core/Card';
import Button from '@material-ui/core/Button';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';

const TreeModule = (props) => {
  const useStyles = makeStyles({
    button: {
      'margin-top': 'auto',
      width: '100%',
    }
  });

  const classes = useStyles();

  return (
    <div className="module-card card darken-1" style={{ height: '100%', width: '100%', display: 'flex', 'flexDirection': 'column' }}>
      <div className="highlight centered">
        <div className="card-body">
          <div className="card-title" style={{ color: '#3F51B5' }}>{props.name}</div>
          <p><strong>Count: </strong>{`${(props.count !== 0) ? props.count : 0}`}</p>
          <p><strong>Percentage: </strong>{`${Math.round(props.count / props.total * 100)}%`}</p>
        </div>
      </div>
      <Button className={classes.button} onClick={props.onClick}>
        {props.button}
      </Button>
    </div>
  )
}

export default TreeModule;