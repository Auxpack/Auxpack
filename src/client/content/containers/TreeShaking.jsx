import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import Typography from '@material-ui/core/Typography';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import { Container } from '@material-ui/core';

// makeStyles used to style Material UI components
const useStyles = makeStyles(theme => ({
  root: {
    width: '100%',
  },
  heading: {
    fontSize: theme.typography.pxToRem(15),
    fontWeight: theme.typography.fontWeightRegular,
  }
}));

const TreeShaking = ({ build, activeBuild }) => {
  if (build[activeBuild].treeStats) {
    // parsing through data for treeshaking information
    const modules = build[activeBuild].treeStats;
    const esmCount = modules.esm.length;
    const cjsCount = modules.cjs.length;
    const bothCount = modules.both.length;
    const totalCount = modules.cjs.length + modules.esm.length + modules.both.length;

    // list items displayed when expansion panel/card opens
    const esmList = modules.esm.map((obj, i) => {
      return (
        <tr key={i} className="table-row">
          <td>{obj.name}</td>
        </tr>
      )
    });

    const cjsList = modules.cjs.map((obj, i) => {
      return (
        <tr key={i} className="table-row">
          <td>{obj.name}</td>
        </tr>
      )
    });

    const bothList = modules.both.map((obj, i) => {
      return (
        <tr key={i} className="table-row">
          <td>{obj.name}</td>
        </tr>
      )
    });

    const totalList = modules.esm.concat(modules.cjs).concat(modules.both).map((obj, i) => {
      return (
        <tr key={i} className="table-row">
          <td>{obj.name}</td>
        </tr>
      )
    })

    // expansion panels/cards
    const TotalCard = () => {
      return (
        <table className="highlight">
          <tbody>
            {totalList}
          </tbody>
        </table>
      )
    }
    const ESMCard = () => {
      return (
        <table className="highlight">
          <tbody>
            {esmList}
          </tbody>
        </table>
      )
    }

    const CJSCard = () => {
      return (
        <table className="highlight">
          <tbody>
            {cjsList}
          </tbody>
        </table>
      )
    }

    const BothCard = () => {
      return (
        <table className="highlight">
          <tbody>
            {bothList}
          </tbody>
        </table>
      )
    }

    const SimpleExpansionPanel = () => {
      const classes = useStyles();

      return (
        <div className='content-card'>
          <div className={classes.root}>
            <ExpansionPanel className="expansionPanel">
              <ExpansionPanelSummary
                expandIcon={<ExpandMoreIcon />}
                aria-controls="panel1a-content"
                id="panel1a-header"
              >
                <Typography className={classes.heading} className="expansion-heading">
                  <strong className="centered">Total Modules</strong>
                  <span>
                    <strong>Count: </strong>{`${(totalCount !== 0) ? totalCount : 0}   `}
                    <strong>Percentage: </strong>{`${Math.round(totalCount / totalCount * 100)}%`}
                  </span>
                </Typography>
              </ExpansionPanelSummary>
              <ExpansionPanelDetails className="panelDetails">
                <TotalCard />
              </ExpansionPanelDetails>
            </ExpansionPanel>
            <ExpansionPanel className="expansionPanel">
              <ExpansionPanelSummary
                expandIcon={<ExpandMoreIcon />}
                aria-controls="panel1a-content"
                id="panel1a-header"
              >
                <Typography className={classes.heading} className="expansion-heading">
                  <strong className="centered">{`Treeshakable (ESM) Modules`}</strong>
                  <span>
                    <strong>Count: </strong>{`${(esmCount !== 0) ? esmCount : 0}   `}
                    <strong>Percentage: </strong>{`${Math.round(esmCount / totalCount * 100)}%`}
                  </span>
                </Typography>
              </ExpansionPanelSummary>
              <ExpansionPanelDetails className="panelDetails">
                <ESMCard />
              </ExpansionPanelDetails>
            </ExpansionPanel>
            <ExpansionPanel className="expansionPanel">
              <ExpansionPanelSummary
                expandIcon={<ExpandMoreIcon />}
                aria-controls="panel1a-content"
                id="panel1a-header"
              >
                <Typography className={classes.heading} className="expansion-heading">
                  <strong className="centered">{`Non-Treeshakable (CJS) Modules`}</strong>
                  <span>
                    <strong>Count: </strong>{`${(cjsCount !== 0) ? cjsCount : 0}   `}
                    <strong>Percentage: </strong>{`${Math.round(cjsCount / totalCount * 100)}%`}
                  </span>
                </Typography>
              </ExpansionPanelSummary>
              <ExpansionPanelDetails className="panelDetails">
                <CJSCard />
              </ExpansionPanelDetails>
            </ExpansionPanel>
            <ExpansionPanel className="expansionPanel">
              <ExpansionPanelSummary
                expandIcon={<ExpandMoreIcon />}
                aria-controls="panel1a-content"
                id="panel1a-header"
              >
                <Typography className={classes.heading} className="expansion-heading">
                  <strong className="centered">{`Mixed Modules`}</strong>
                  <span>
                    <strong>Count: </strong>{`${(bothCount !== 0) ? bothCount : 0}   `}
                    <strong>Percentage: </strong>{`${Math.round(bothCount / totalCount * 100)}%`}
                  </span>
                </Typography>
              </ExpansionPanelSummary>
              <ExpansionPanelDetails className="panelDetails">
                <BothCard />
              </ExpansionPanelDetails>
            </ExpansionPanel>
          </div>
        </div>
      )
    }

    return (
      <div className="cards-container centered">
        <SimpleExpansionPanel />
      </div>
    )
  } else {
    return (<div className="centered cards-container">
      <p>No treeshakable modules available.</p>
    </div>)
  }
}

export default TreeShaking;




