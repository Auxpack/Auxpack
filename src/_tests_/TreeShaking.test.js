import TreeShaking from '../client/content/containers/TreeShaking.jsx';
import TreeModule from '../client/content/components/TreeModule.jsx';
import ExpansionPanel from '@material-ui/core/ExpansionPanel';

describe('TreeShaking Unit Tests', () => {

  describe('TreeShaking Container', () => {
    let shallow;
    let wrapper;

    const props = {
      build:[{
        timeStamp:1575426090404,
        time:1439,
        hash:"546142ce1b49a6ba7d6f",
        errors:[],
        size:1172113,
        assets:[{"name":"bundle.js","chunks":["main"],"size":1172113}],
        chunks:[{"size":1118609,"files":["bundle.js"],"modules":[{"name":"./client/App.jsx","size":6375,"id":"./client/App.jsx"}]}],
        treeStats:{
          "cjs":[{"name":"./test2.js","size":49,"reasonTypes":[{"name":"./test.js","type":"cjs require"}]},{"name":"./test2.js","size":49,"reasonTypes":[{"name":"./test.js","type":"cjs require"}]}],
          "esm":[{"name":"./client/App.jsx","size":6375,"reasonTypes":[{"name":"./client/index.js","type":"harmony side effect evaluation"}]}],
          "both":[{"name":"./node_modules/react/index.js","size":190,"reasonTypes":[{"name":"./client/App.jsx","type":"harmony side effect evaluation"}]}]}
        }],
      activeBuild: 0
    };

    beforeAll(() => {
      shallow = createShallow();
      wrapper = shallow(<TreeShaking {...props} />);
    });

    it('TreeShaking snapshot testing', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('TreeShaking should render four TreeModule components', () => {
      expect((wrapper.find(TreeModule)).length).toBe(4);
    });

    it('TreeShaking should display correct module counts inside each TreeModule component', () => {
      // expect total count to be sum of lengths of props.build[0].treeStats
      wrapper.find(TreeModule).forEach(node => {
        expect(node.prop('totalCount')).toBe(4)
      });
      // expect count displayed in module 1 to be equal to total count
      expect(wrapper.find(TreeModule).at(0).prop('count')).toBe(4);
      // expect count displayed in module 2 to be equal to esm count
      expect(wrapper.find(TreeModule).at(1).prop('count')).toBe(1);
      // expect count displayed in module 3 to be equal to cjs count
      expect(wrapper.find(TreeModule).at(2).prop('count')).toBe(2);
      // expect count displayed in module 4 to be equal to both count
      expect(wrapper.find(TreeModule).at(3).prop('count')).toBe(1);
    });
  });

  describe('TreeModule', () => {
    let shallow;
    let wrapper;
    const props = {
      list: [{name:'one'}, {name:'two'}, {name:'three'}], 
      title: 'total', 
      count: 5, 
      totalCount: 10
    };

    beforeAll(() => {
      shallow = createShallow();
      wrapper = shallow(<TreeModule {...props} />);
    });

    it('TreeModule snapshot testing', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('TreeModule should render an Expansion Panel element', () => {
      expect((wrapper.find(ExpansionPanel)).length).toBe(1);
    });
  });
});
