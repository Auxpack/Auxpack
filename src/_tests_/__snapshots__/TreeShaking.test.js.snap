// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeShaking Unit Tests TreeModule TreeModule snapshot testing 1`] = `ShallowWrapper {}`;

exports[`TreeShaking Unit Tests TreeShaking Container TreeShaking snapshot testing 1`] = `ShallowWrapper {}`;
